rust/kernel/net/mod.rs

```diff
@@ -0,0 +1,108 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Network subsystem in rust for linux kernel.
+
+use core::mem;
+
+pub mod device;
+pub mod ethtool;
+pub mod netlink;
+pub mod rtnl;
+pub mod skbuff;
+
+#[doc(inline)]
+pub use macros::rtnl_link_ops;
+
+/// Determine if the Ethernet address is a multicast.
+///
+/// Return true if the address is a multicast address.
+/// By definition the broadcast address is also a multicast address.
+///
+/// # Parameters
+/// - `addr`: Pointer to a six-byte array containing the Ethernet address
+///
+/// # Safety
+/// `addr` has to be a valid address pointer
+#[cfg(CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS)]
+pub unsafe fn is_multicast_ether_addr(addr: *const u8) -> bool {
+    let a: u32 = unsafe { *(addr as *const u32) };
+
+    if cfg!(target_endian = "big") {
+        (0x01 & (a >> (((mem::size_of::<u32>() as u32) * 8) - 8))) != 0
+    } else {
+        (0x01 & a) != 0
+    }
+}
+
+/// Determine if the Ethernet address is a multicast.
+///
+/// Return true if the address is a multicast address.
+/// By definition the broadcast address is also a multicast address.
+///
+/// # Parameters
+/// - `addr`: Pointer to a six-byte array containing the Ethernet address
+///
+/// # Safety
+/// `addr` has to be a valid address pointer
+#[cfg(not(CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS))]
+pub unsafe fn is_multicast_ether_addr(addr: *const u8) -> bool {
+    let a: u16 = unsafe { *(addr as *const u16) };
+
+    if cfg!(target_endian = "big") {
+        (0x01 & (a >> (((mem::size_of::<u16>() as u16) * 8) - 8))) != 0
+    } else {
+        (0x01 & a) != 0
+    }
+}
+
+/// Determine if give Ethernet address is all zeros.
+/// Return true if the address is all zeroes.
+///
+/// # Parameters
+/// - `addr`: Pointer to a six-byte array containing the Ethernet address
+///
+/// # Safety
+/// Please note: addr must be aligned to u16.
+#[cfg(CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS)]
+pub unsafe fn is_zero_ether_addr(addr: *const u8) -> bool {
+    // SAFETY: function already unsafe
```

Comment:
This is not correct. You should explain why the function is unsafe on the safety requirement above, e.g. say "`addr` must point to a valid ethernet address and 16-bit aligned, and then use the safety requirement to justify why the function is safe.

Reply:
Is this better now?

Reply:
I wonder if an encapsulation for a known-to-be-valid ethernet address pointer would help, so that we don't need these unsafe-to-call functions.

Or maybe we should just bind directly to C implementations via helpers for now, and we can come back for performance optimisations later.

Reply:
Yes I would like an Wrapper for that in the long run. for now I think it's a bit much, It's already +4k. Keep it like this until I probably need to change it for e1000?